File: solutions/3755-maximum-product-of-first-and-last-elements-of-a-subsequence/maximum-product-of-first-and-last-elements-of-a-subsequence.java
class Solution {
    public long maximumProduct(int[] nums, int m) {
        int n = nums.length;
        int[] max = new int[n];
        int[] min = new int[n];
        max[n-1]=min[n-1]=nums[n-1];
        for(int i=n-2;i>=0;i--){
            max[i] = Math.max(max[i+1],nums[i]);
            min[i] = Math.min(min[i+1],nums[i]);
        }

        long ans = Long.MIN_VALUE;

        for(int i=0;i<n;i++){
            int j = i+m-1;
            if(j>=n) break;
            long product = 1L*nums[i]*(nums[i]>=0?max[j]:min[j]);
            ans = Math.max(ans,product);
        }

        return ans;
    }
}

File: solutions/3755-maximum-product-of-first-and-last-elements-of-a-subsequence/maximum-product-of-first-and-last-elements-of-a-subsequence.cpp
typedef long long ll;
class Solution {
public:
    ll maximumProduct(vector<int>& nums, int m) {
        int n = nums.size();
        vector<ll> rmax(n, INT_MIN), rmin(n, INT_MAX);
        vector<ll> lmax(n, INT_MIN), lmin(n, INT_MAX);
        
        rmax[n-1] = nums[n-1];
        rmin[n-1] = nums[n-1];
        lmax[0] = nums[0];
        lmin[0] = nums[0];

        // Build suffix max/min
        for (int i = n-2; i >= 0; i--) {
            rmax[i] = max(rmax[i+1], (ll)nums[i]);
            rmin[i] = min(rmin[i+1], (ll)nums[i]);
        }

        ll res = LLONG_MIN;
        for (int i = 0; i < n; i++) {
            if (i > 0) {
                lmax[i] = max(lmax[i-1], (ll)nums[i]);
                lmin[i] = min(lmin[i-1], (ll)nums[i]);
            }

            ll previ = i - m + 1;
            if (previ >= 0) {
                res = max(res, rmax[i] * lmax[previ]);
                res = max(res, rmin[i] * lmin[previ]);
            }
        }
        return res;
    }
};